nsfc  plugin register, update,delete and get networkelements for sfc

karimull opened the pull request.

arvindn05 commented on src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java:
```diff
@@ -246,25 +248,84 @@ public InspectionHookElement getInspectionHook(String inspectionHookId) throws E
+       
+        ServiceFunctionChainEntity sfc = new ServiceFunctionChainEntity();
+        //check for null or empty list
+        this.utils.throwExceptionIfNullOrEmptyNetworElementList(inspectionPorts, "PortPairGroup memeber list");
```
spaces between PortPairGroup

karimull replied:
modified

arvindn05 commented on src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java:
```diff
@@ -246,25 +248,84 @@ public InspectionHookElement getInspectionHook(String inspectionHookId) throws E
-        // no-op
-        return null;
+        
+        this.utils.throwExceptionIfNullElementAndParentId(portGroup, "PortPairGroup ServiceFunctionChain Id");
```
why do we check parent id? we should just check element id. NetworkElement portGroup -> is actually sfc

karimull replied:
Was expecting sfc_id in parent...modified to element-id now.

arvindn05 commented on src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java:
```diff
@@ -246,25 +248,84 @@ public InspectionHookElement getInspectionHook(String inspectionHookId) throws E
+        this.utils.throwExceptionIfNullOrEmptyNetworElementList(inspectionPorts, "PortPairGroup update memeber list");
+       
+        return this.txControl.required(() -> {
+            ServiceFunctionChainEntity sfc = this.utils.findBySfcId(portGroup.getParentId());
```
please do this outside of the transaction

karimull replied:
Modified for all APIs where needed moved outside the transaction

arvindn05 commented on src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java:
```diff
@@ -246,25 +248,84 @@ public InspectionHookElement getInspectionHook(String inspectionHookId) throws E
     public void deleteNetworkElement(NetworkElement portGroupId) throws Exception {
-        // no-op
+        this.utils.throwExceptionIfNullElementAndId(portGroupId, "PortPairGroupId");
+        this.utils.throwExceptionIfNullElementAndParentId(portGroupId, "PortPairGroupId Parent");
```
we dont care about parent id. here again we are getting sfc directly

karimull replied:
This is delete network element. Which is we are deleting PPG from sfc. So network element will have element-id as PPG-id and parent-id will be the sfc id. So if see any of them null we have to throw an exception. Please let me know if i'm missing anything here.

arvindn05 replied:
delete network element deletes the SFC. Similar to how register network elements creates a sfc.

To delete a PPG from an SFC, you can use the updateNetworkElements

karimull replied:
got it..fixed thanks

karimull commented:
rebased it..now should be able to see utils file changes.

karimull commented:
Added unit tests and fixed following issues
1. Return of network element after register and update
2. Delete of sfc from data base.
3. Few checks fir element and parent id.
4. Modified log messages.

arvindn05 commented on src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java:
```diff
@@ -246,25 +248,89 @@ public InspectionHookElement getInspectionHook(String inspectionHookId) throws E
-        // no-op
-        return null;
+
+        this.utils.throwExceptionIfNullElementAndId(portGroup, "Port Pair Group ServiceFunctionChain Id");
```
add space between sfc

karimull replied:
corrected

arvindn05 commented on src/main/java/org/osc/controller/nsfc/utils/RedirectionApiUtils.java:
```diff
@@ -222,7 +223,28 @@ public InspectionHookEntity findInspHookByInspectedAndPort(NetworkElement inspec
         });
     }
+    
+    public void addPortPairGroupToServiceFunction(List<NetworkElement> inspectionPorts,
```
rename inspectionPorts to portPairgroups so it reads easier.

karimull replied:
corrected

arvindn05 commented on src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java:
```diff
@@ -246,25 +248,89 @@ public InspectionHookElement getInspectionHook(String inspectionHookId) throws E
+        ServiceFunctionChainEntity sfc = this.utils.findBySfcId(portGroup.getElementId());
+
+        if (sfc == null) {
+            throw new IllegalStateException(String.format("Unable to find ServiceFunctionChain Id : %s  Update failed",
```
use throwExceptionIfCannotFindById

karimull replied:
modified..in call findbyId calls

arvindn05 commented on src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java:
```diff
@@ -246,25 +248,89 @@ public InspectionHookElement getInspectionHook(String inspectionHookId) throws E
+
+        ServiceFunctionChainEntity sfc = this.utils.findBySfcId(serviceFunctionChainId.getElementId());
+        if (sfc == null) {
+            throw new IllegalStateException(String.format("Unable to find ServiceFunctionChain Id : %s  Delete failed",
```
use throwExceptionIfCannotFindById

arvindn05 commented on src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java:
```diff
@@ -246,25 +248,89 @@ public InspectionHookElement getInspectionHook(String inspectionHookId) throws E
+
+        this.txControl.required(() -> {
+            //need to delete sfc if found in a transaction so fetch it again
+            ServiceFunctionChainEntity sfcToDel = this.em.find(ServiceFunctionChainEntity.class, serviceFunctionChainId.getElementId());
```
move the findBySfcId method here

karimull replied:
The unit test is failing with exception if we are throwing a exception in transaction.
So I have to opt for a check before entering into the transaction. Please Let me know if there is another way.

arvindn05 replied:
hmm...we are throwing in transaction in other places....what  is different here? lets discuss tommorow.

karimull replied:
used try, catch and it seems to be working. Please check and let me know.
we can throw an exception using a msg or I can call the same util API to throw the exception. Which were is preferred..

arvindn05 replied:
i've tested locally. Exceptions thrown within transactions are wrapped with ScopedWorkException looks like...
So you should catch this exception and unwrap if required

arvindn05 commented on src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java:
```diff
@@ -246,25 +248,89 @@ public InspectionHookElement getInspectionHook(String inspectionHookId) throws E
+                    element.getElementId()));
+        }
+        //do we need to have transaction for get?
+        return this.txControl.required(() -> {
```
you can just return sfc.getPortPairGroups since ppg's are network elements

karimull replied:
Couldn't modify as it is giving a compilation error.. 
return sfc.getPortPairGroups() -- giving a compile error

arvindn05 replied:
try new ArrayList<>(sfc.getPortPairGroups())

karimull replied:
Thanks that worked..fixed it

karimull commented on src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java:
```diff
@@ -246,25 +248,89 @@ public InspectionHookElement getInspectionHook(String inspectionHookId) throws E
+
+        ServiceFunctionChainEntity sfc = new ServiceFunctionChainEntity();
+        //check for null or empty list
+        this.utils.throwExceptionIfNullOrEmptyNetworElementList(inspectionPorts, "Port Pair Group memeber list");
```
corrected

karimull commented on src/main/java/org/osc/controller/nsfc/utils/RedirectionApiUtils.java:
```diff
@@ -222,7 +223,28 @@ public InspectionHookEntity findInspHookByInspectedAndPort(NetworkElement inspec
         });
     }
+    
+    public void addPortPairGroupToServiceFunction(List<NetworkElement> inspectionPorts,
```
modified

karimull commented on src/main/java/org/osc/controller/nsfc/utils/RedirectionApiUtils.java:
```diff
@@ -222,7 +223,28 @@ public InspectionHookEntity findInspHookByInspectedAndPort(NetworkElement inspec
+        List<PortPairGroupEntity> ppgList = new ArrayList<PortPairGroupEntity>();
+        for (NetworkElement ne : inspectionPorts) {
+            throwExceptionIfNullElementAndId(ne, "Port Pair Group Id");
+            if(sfc.getElementId() != null) { //then we should have parent element
```
removed the check..as we are expecting sfc id in element id now.

karimull commented on src/main/java/org/osc/controller/nsfc/utils/RedirectionApiUtils.java:
```diff
@@ -222,7 +223,28 @@ public InspectionHookEntity findInspHookByInspectedAndPort(NetworkElement inspec
+                throwExceptionIfNullElementAndParentId(ne, "Port Pair Group ServiceFunctionChain Id"); 
+            }
 
+            if (findByPortPairgroupId(ne.getElementId()) == null) {
```
corrected
